src/urlpattern_pattern_parser.cpp

```diff
@@ -0,0 +1,339 @@
+      modifier == PART_MODIFIER::NONE) {
+    // This was a "{foo}" grouping. We add this to the pending fixed value so
+    // that it will be combined with any previous or subsequent text.
+    // 1. Append prefix to the end of parser’s pending fixed value.
+    pending_fixed_value.append(prefix);
+    return;
+  }
+  // 4. Run maybe add a part from the pending fixed value given parser.
+  maybe_add_part_from_pendind_fixed_value();
+  // 5. If name token is null and regexp or wildcard token is null:
+  if (!name_token.has_value() && !regexp_or_wildcard_token.has_value()) {
+    // This was a "{foo}?" grouping. The modifier means we cannot combine it
+    // with other text. Therefore we add it as a part immediately.
+    // 1. Assert: suffix is the empty string.
+    assert(suffix.empty());
+    // 2. If prefix is the empty string, then return.
+    if (prefix.empty()) return;
+    // 3.Let encoded value be the result of running parser’s encoding callback
+    // given prefix.
+    auto encoded_value = encoding_callback(prefix);
+
+    // 4. Let part be a new part whose type is "fixed-text", value is encoded
+    // value, and modifier is modifier.
+    auto p = part();
+    p.type = PART_TYPE::FIXED_TEXT;
+    p.value = encoded_value;
+    p.modifier = modifier;
+
+    // 5. Append part to parser’s part list.
+    part_list.push_back(p);
+    return;
+  }
+
+  // 6. Let regexp value be the empty string.
+  std::u32string regexp_value{};
+
+  // Next, we convert the regexp or wildcard token into a regular expression.
+
+  // 7. If regexp or wildcard token is null, then set regexp value to parser’s
+  // segment wildcard regexp.
+  if (!regexp_or_wildcard_token.has_value()) {
+    regexp_value = segment_wildcard_regexp;
+  }
+  // 8. Else if regexp or wildcard token’s type is "asterisk", then set regexp
+  // value to the full wildcard regexp value.
+  else if (regexp_or_wildcard_token.value()->type == TOKEN_TYPE::ASTERISK) {
+    regexp_value = full_wildcard_regexp_value;
+  }
+  // 9. Else set regexp value to regexp or wildcard token’s value.
+  else {
+    regexp_value = regexp_or_wildcard_token.value()->value;
+  }
+  // 10. Let type be "regexp".
+  auto type = PART_TYPE::REGEXP;
+  // Next, we convert regexp value into a part type. We make sure to go to a
+  // regular expression first so that an equivalent "regexp" token will be
+  // treated the same as a "name" or "asterisk" token.
+
+  // 11. If regexp value is parser’s segment wildcard regexp:
+  if (regexp_value == segment_wildcard_regexp) {
+    // 1. Set type to "segment-wildcard".
+    type = PART_TYPE::SEGMENT_WILDCARD;
+    // 2. Set regexp value to the empty string.
+    regexp_value.clear();
+  }
+  // 12. Else if regexp value is the full wildcard regexp value:
+  else if (regexp_value == full_wildcard_regexp_value) {
+    // 1. Set type to "full-wildcard".
+    type = PART_TYPE::FULL_WILDCARD;
+    // 2. Set regexp value to the empty string.
+    regexp_value.clear();
```

Comment:
We are assigning and later clearing this value, which is creating performance degregation. We should eventually optimize it to reduce unnecessary allocations.